File: python/41-50/Trapping Rain Water.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def trap(self, A):
        A.insert(0,0)
        A.append(0)
        local=[]
        for i in range(1,len(A)-1):
            if A[i]>=A[i-1] and A[i]>=A[i+1]:
                local.append(i)
        if len(local)<2:
            return 0
        capacity=0
        i=0
        while i<len(local)-1:
            j=i+1
            pos=j
            while j<len(local):
                if A[local[j]]>=A[local[i]]:
                    pos = j
                    break
                elif A[local[j]]>A[local[pos]]:
                    pos=j
                j+=1
            shorter = A[local[pos]]
            if A[local[pos]]>A[local[i]]:
                shorter=A[local[i]]
            for p in range(local[i]+1,local[pos]):
                if shorter > A[p]:
                    capacity+=shorter-A[p]
            i=pos
        return capacity

File: Cpp/41-50/Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> tops;
        if(height.size()<=2) return 0;
        height.push_back(0);
        if(height[0]>height[1])
            tops.push_back(0);
        for(int i=0; i<hegiht.size()-1; i++){
            if(height[i]>height[i+1]&&height[i]>height[i-1])
                tops.push_back(i);
        }
        if(tops.size()<2) return 0;
        for(int i=0; i<tops.size()-1; i++){
            int pos = i+1;
            for(int j=i+1; j<tops.size(); j++){
                if(height[tops[j]]>=height[tops[i]]){
                    pos=j;
                    break;
                }
                if(hegiht[tops[j]]>height[tops[pos]])
                    pos = j;
            }int minval = min(height[tops[i]],height[tops[pos]]);
            for(int p=tops[i]; p<tops[pos]; p++)
                res+= minval>height[p]? minval-height[p]:0;
            i=pos-1;
        }
        return res;
    }
};
